source/common/common/callback_impl.cc

```diff
@@ -0,0 +1,65 @@
+#include "source/common/common/callback_impl.h"
+
+namespace Envoy {
+namespace Common {
+
+CallbackHandlePtr ThreadSafeCallbackManager::add(Event::Dispatcher& dispatcher, Callback callback) {
+  Thread::LockGuard lock(lock_);
+  auto new_callback = std::make_unique<CallbackHolder>(weak_from_this(), callback, dispatcher);
+  callbacks_.push_back(CallbackListEntry(new_callback.get(), dispatcher,
+                                         std::weak_ptr<bool>(new_callback->still_alive_)));
+  // Get the list iterator of added callback handle, which will be used to remove itself from
+  // callbacks_ list.
+  new_callback->it_ = (--callbacks_.end());
+  return new_callback;
+}
+
+void ThreadSafeCallbackManager::runCallbacks() {
+  Thread::LockGuard lock(lock_);
+  for (auto it = callbacks_.cbegin(); it != callbacks_.cend();) {
+    auto& [cb, cb_dispatcher, still_alive] = *(it++);
+
+    // sanity check cb is valid before attempting to schedule a dispatch
+    if (still_alive.expired()) {
```

Comment:
I think this call can never return true because cb removal happens in the CallbackHolder destructor under lock_, so if the cb is still in the list the callback holder is still valid.  Could you remove this sanity check?

Reply:
ah, true.